File: games/backgammon-unity/Assets/Scripts/BackgammonState.cs
using System.Collections.Generic;

[System.Serializable]
public enum BackgammonColor
{
    White,
    Black
}

[System.Serializable]
public class Point
{
    public int index;
    public List<BackgammonColor> checkers = new List<BackgammonColor>();
    public bool IsBlot => checkers.Count == 1;
    public bool IsBlocked => checkers.Count >= 2;
    public BackgammonColor? Owner => checkers.Count > 0 ? checkers[0] : null;

    public Point(int idx)
    {
        index = idx;
    }
}

public class BackgammonState
{
    public List<string> players;
    public string currentPlayer;
    public BackgammonColor turn = BackgammonColor.White;
    public List<Point> points;
    public int barWhite = 0;
    public int barBlack = 0;
    public int borneOffWhite = 0;
    public int borneOffBlack = 0;
    public List<int> dice = new List<int>();
    public bool diceRolled = false;
    public bool gameOver = false;

    public BackgammonState(List<string> playerIds)
    {
        players = playerIds;
        currentPlayer = playerIds[0];
        points = CreateInitialPoints();
    }

    public static List<Point> CreateInitialPoints()
    {
        var points = new List<Point>();
        for (int i = 0; i < 25; i++)
        {
            points.Add(new Point(i));
        }

        // White's checkers
        points[24].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.White, BackgammonColor.White });
        points[13].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.White, BackgammonColor.White, BackgammonColor.White, BackgammonColor.White, BackgammonColor.White });
        points[8].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.White, BackgammonColor.White, BackgammonColor.White });
        points[6].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.White, BackgammonColor.White, BackgammonColor.White, BackgammonColor.White, BackgammonColor.White });

        // Black's checkers
        points[1].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.Black, BackgammonColor.Black });
        points[12].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black });
        points[17].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black });
        points[19].checkers.AddRange(new List<BackgammonColor> { BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black, BackgammonColor.Black });

        return points;
    }
}

File: games/backgammon/src/commonMain/kotlin/com/gamehub/games/backgammon/BackgammonState.kt
package com.gamehub.games.backgammon

import com.gamehub.shared.core.GameState
import com.gamehub.shared.core.PlayerId
import kotlinx.serialization.Serializable
import com.gamehub.shared.dice.DiceEngine
import kotlinx.serialization.Transient

@Serializable
enum class BackgammonColor {
    WHITE,
    BLACK
}

@Serializable
data class Point(
    val index: Int, // 1-24
    val checkers: List<BackgammonColor> = emptyList()
) {
    val isBlot: Boolean
        get() = checkers.size == 1
    val isBlocked: Boolean
        get() = checkers.size >= 2
    val owner: BackgammonColor?
        get() = checkers.firstOrNull()

    fun addChecker(color: BackgammonColor): Point {
        return copy(checkers = checkers + color)
    }

    fun removeLastChecker(): Point {
        return copy(checkers = checkers.dropLast(1))
    }

    fun clearCheckers(): Point {
        return copy(checkers = emptyList())
    }
}

@Serializable
data class DoublingCube(
    val value: Int = 1, // 1, 2, 4, 8, 16, 32, 64
    val owner: PlayerId? = null // null means center
)

@Serializable
data class BackgammonState(
    val players: List<PlayerId>,
    val currentPlayer: PlayerId?,
    val turn: BackgammonColor = BackgammonColor.WHITE,
    val points: List<Point> = createInitialPoints(),
    val barWhite: Int = 0,
    val barBlack: Int = 0,
    val borneOffWhite: Int = 0,
    val borneOffBlack: Int = 0,
    val dice: List<Int> = emptyList(),
    val diceRolled: Boolean = false,
    val doublingCube: DoublingCube = DoublingCube(),
    val canOfferDouble: Boolean = true,
    val gameOver: Boolean = false,
    @Transient
    val diceEngines: Map<PlayerId, DiceEngine> = emptyMap() // موتور تاس هر بازیکن
) : GameState() {
    companion object {
        fun createInitialPoints(): List<Point> {
            val points = MutableList(25) { idx -> Point(idx) } // 0 unused, 1-24

            // From user's image:
            points[24] = points[24].copy(checkers = List(2) { BackgammonColor.BLACK })
            points[13] = points[13].copy(checkers = List(5) { BackgammonColor.BLACK })
            points[8] = points[8].copy(checkers = List(3) { BackgammonColor.BLACK })
            points[6] = points[6].copy(checkers = List(5) { BackgammonColor.BLACK })
            points[1] = points[1].copy(checkers = List(2) { BackgammonColor.WHITE })
            points[19] = points[19].copy(checkers = List(5) { BackgammonColor.WHITE })
            points[17] = points[17].copy(checkers = List(3) { BackgammonColor.WHITE })
            points[12] = points[12].copy(checkers = List(5) { BackgammonColor.WHITE })

            return points
        }
    }

    fun getPlayerColor(playerId: PlayerId): BackgammonColor {
        return if (players.indexOf(playerId) == 0) BackgammonColor.WHITE else BackgammonColor.BLACK
    }
}
